dataService.js

```diff
@@ -0,0 +1,71 @@
+// Required Maps
+let productMap;
+let categoryMap;
+let combinedProductMap;
+
+/*
+ * Method to create and return map of products if its not already created
+ */
+function getProducts() {
+
+  if (!productMap || !productMap.size) {
+    const productData = require('./src/data/products.json');
+    productMap = {};
+
+    productData.products.forEach(product => {
+      productMap[product.id] = product;
+    });
+  }
+
+  return productMap;
+}
+
+/*
+ * Method to create and return map of categories if not already created
+ */
+function getCategories() {
+
+  if (!categoryMap || !categoryMap.size) {
+    const categoryData = require('./src/data/categories.json');
+    categoryMap = {};
+
+    categoryData.categories.forEach(category => {
+      categoryMap[category.id] = category.categoryName;
+    });
+  }
+
+  return categoryMap;
+}
+
+/*
+ * Method to combine product and category data into combinedProductMap
+ */
+function combineProductsWithCategories() {
+
+  // Creating productMap and categoryMap using above functions
+  getProducts();
+  getCategories();
+
+  combinedProductMap = {};
+
+  Object.values(productMap).forEach(product => {
+    product.categoryName = categoryMap[product.categoryId];
+    combinedProductMap[product.id] = product;
+  });
+}
+
+/*
+ * Creates combinedProductMap if it is null
+ */
+function getCombinedProductMap() {
+
+  if (!combinedProductMap || !combinedProductMap.size) {
+    combineProductsWithCategories();
+  }
+  return combinedProductMap;
+}
+
+module.exports = {
+  getCategories: getCategories(),
```

Comment:
Can't we simply export `getCategories` ?

Reply:
I didn't get that? Do you mean exporting the variable?

Reply:
Ok, I think I got it.

Reply:
I tried simplifying it by using arrow functions but its giving run time errors, I think we're good to go with `getCategories()`.

Same for `getCombinedProductMap()`